hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/AbstractReactiveSaveEventListener.java

```diff
@@ -144,55 +144,58 @@ else if ( generator instanceof Assigned ) {
-				return ( (ReactiveIdentifierGenerator<?>) generator )
-						.generate( ( ReactiveConnectionSupplier ) source, entity )
-						.thenApply( id -> castToIdentifierType( id, persister ) )
-						.thenCompose( gid -> performSaveWithId(
-								entity,
-								context,
-								source,
-								persister,
-								generator,
-								gid,
-								requiresImmediateIdAccess,
-								false
-						) );
+				return generateId( entity, source, (ReactiveIdentifierGenerator<?>) generator, persister )
+						.thenCompose( gid -> {
+							if ( gid == SHORT_CIRCUIT_INDICATOR ) {
+								source.getIdentifier( entity );
+								return voidFuture();
+							}
+							persister.setIdentifier( entity, gid, source );
+							return reactivePerformSave(
+									entity,
+									gid,
+									persister,
+									generatedOnExecution,
+									context,
+									source,
+									false
+							);
+						} );
 			}
 
 			generatedId = ( (BeforeExecutionGenerator) generator ).generate( source, entity, null, INSERT );
+			if ( generatedId == SHORT_CIRCUIT_INDICATOR ) {
+				source.getIdentifier( entity );
+				return voidFuture();
+			}
+			persister.setIdentifier( entity, generatedId, source );
 		}
 		final Object id =  castToIdentifierType( generatedId, persister );
-		return reactivePerformSave( entity, id, persister, generatedOnExecution, context, source, requiresImmediateIdAccess );
+		final boolean delayIdentityInserts = !source.isTransactionInProgress() && !requiresImmediateIdAccess && generatedOnExecution;
+		return reactivePerformSave( entity, id, persister, generatedOnExecution, context, source, delayIdentityInserts );
 	}
 
-	private CompletionStage<Void> performSaveWithId(
+	private CompletionStage<Object> generateId(
 			Object entity,
-			C context,
 			EventSource source,
-			EntityPersister persister,
-			Generator generator,
-			Object generatedId,
-			boolean requiresImmediateIdAccess,
-			boolean generatedOnExecution) {
-		if ( generatedId == null ) {
-			throw new IdentifierGenerationException( "null id generated for: " + entity.getClass() );
-		}
-		if ( generatedId == SHORT_CIRCUIT_INDICATOR ) {
-			source.getIdentifier( entity );
-			return voidFuture();
-		}
-		if ( LOG.isDebugEnabled() ) {
-			LOG.debugf(
-					"Generated identifier: %s, using strategy: %s",
-					persister.getIdentifierType().toLoggableString( generatedId, source.getFactory() ),
-					generator.getClass().getName()
-			);
-		}
-		final boolean delayIdentityInserts =
-				!source.isTransactionInProgress()
-						&& !requiresImmediateIdAccess
-						&& generatedOnExecution;
-		return reactivePerformSave( entity, generatedId, persister, false, context, source, delayIdentityInserts );
+			ReactiveIdentifierGenerator<?> generator,
+			EntityPersister persister) {
+		return generator
+				.generate( (ReactiveConnectionSupplier) source, entity )
+				.thenApply( id -> castToIdentifierType( id, persister ) )
+				.thenCompose( generatedId -> {
```

Comment:
The code works, but it doesn't seem like you have to use  `.thenCompose`.
You could keep everything in the `.thenApply`, this way we avoid the creation of a couple of `CompletionStage` objects:
```java
.thenApply( id -> {
	Object generatedId = castToIdentifierType( id, persister );
	if ( generatedId == null ) {
		throw new IdentifierGenerationException( "null id generated for: " + entity.getClass() );
	}
	if ( LOG.isDebugEnabled() ) {
		LOG.debugf(
				"Generated identifier: %s, using strategy: %s",
				persister.getIdentifierType().toLoggableString( generatedId, source.getFactory() ),
				generator.getClass().getName()
		);
	}
	return generatedId;
} );
```